src/snowflake/cli/plugins/object/stage/diff.py

```diff
@@ -154,7 +162,47 @@ def build_md5_map(list_stage_cursor: SnowflakeCursor) -> Dict[str, str]:
     }
 
 
-def stage_diff(local_path: Path, stage_fqn: str) -> DiffResult:
+def _get_relative_paths_to_sync(
+    relative_files_to_sync: List[Path], local_path: Path, remote_paths: Set[str]
+) -> List[str]:
+    paths = []
+    for file in relative_files_to_sync:
+        path = Path(os.path.join(local_path, file))
+        relpath = path.relative_to(local_path)
+        if not path.exists() and str(relpath) not in remote_paths:
+            raise FileError(
+                str(file), "This file does not exist either locally or remotely"
+            )
+        elif path.is_dir():
+            raise ClickException(f"Specifying directories is not supported: '{file}'")
+        else:
+            paths.append(str(relpath))
+    return paths
+
+
+def _filter_from_diff(
+    result: DiffResult, paths_to_keep: Set[str], prune: bool
+) -> DiffResult:
+    result.different = [i for i in result.different if i in paths_to_keep]
+    result.only_local = [i for i in result.only_local if i in paths_to_keep]
+    if prune:
+        result.only_on_stage = [i for i in result.only_on_stage if i in paths_to_keep]
+    else:
+        files_not_removed = [i for i in result.only_on_stage if i in paths_to_keep]
+        if len(files_not_removed) > 0:
+            cc.warning(
+                f"The following files exist only on the stage:\n{files_not_removed}\nUse the --prune flag to delete them from the stage."
+            )
+        result.only_on_stage = []
+    return result
+
+
+def stage_diff(
+    local_path: Path,
+    stage_fqn: str,
+    prune: bool = True,
+    files_to_sync: Optional[List[Path]] = None,  # relative to local_path
```

Comment:
Didn't we agree to make prune and file filters subsequent passes on the original diff? I don't think this belongs here, let's keep the stage diff logic simple and clean unless there's a really compelling reason to do this here that I'm not seeing.